designs/nutanix-datacenterconfig-credentialref.md

```diff
@@ -0,0 +1,51 @@
+# Nutanix Datacenter Config Credential Reference
+
+## Introduction
+
+**Problem:** When a management cluster and a workload cluster are present on two different Nutanix Prism Central instances,
+using same credentials to authenticate to both instances is not possible. To remedy this, we need to associate credentials
+with corresponding Nutanix Datacenter Configs.
+
+## Overview of Solution
+With this feature, a user can specify credentials reference (i.e. Name of the secret in eksa-system namespace containing the Prism credentials)
+in the Nutanix Datacenter Config. This info will be fetched and used to authenticate to the Nutanix Prism Central instance
+when creating the workload cluster.
+
+### Solution Details
+Extend the `NutanixDatacenterConfig` CRD to include a new field `credentialRef` storing name (name of kubernetes secret object) and
+kind ("Secret"). This is similar to how `credentialsRef` is defined in the `NutanixCluster` CRD minus the namespace. The namespace
+of the secret will always be `eksa-system`. Keeping the namespace constant will allow us to avoid granting EKS-A controller access to
+secrets in other namespaces. The `Ref` type used in Snow provider is reused here for this purpose.
+
+```go
+// NutanixDatacenterConfigSpec defines the desired state of NutanixDatacenterConfig.
+type NutanixDatacenterConfigSpec struct {
+    ...
+    // CredentialRef is the reference to the secret name that contains the credentials
+    // for the Nutanix Prism Central. The namespace for the secret is assumed to be a constant i.e. eksa-system.
+    // +kubebuilder:validation:Optional
+    // +optional
+    CredentialRef *Ref `json:"credentialRef,omitempty"`
+}
+
+func (in *NutanixDatacenterConfig) Validate() error {
+    ...
+    if in.Spec.CredentialRef != nil && in.Spec.CredentialRef.Kind != RefKindSecret {
+        return errors.New("only Secret is supported as a credential kind")
+    }
+    ...
+}
+```
+
+When creating a workload cluster, the `NutanixDatacenterConfig` referenced in the `Cluster` CRD will be fetched and
+the `credentialRef` field will be used to fetch the secret to get the credentials needed to authenticate with Prism Central
+to create resources needed for the cluster. If the `credentialRef` field is not present when creating a nutanix datacenter config, a
+mutation webhook will add the `credentialRef` field to the `NutanixDatacenterConfig` with the `nutanix-credentials` as the
```

Comment:
I thought we said there wasn't going to be any defaulting? The user has to set credentialsRef, and if they don't, the webhook will error out.

For management clusters, since they are created through the CLI, the CLI will take care of filling the credentialsRef and create the secret, pulling the info from the current env vars.

Reply:
Got it; removed the usage of defaulting webhook.